src/or/connection_edge.c

```diff
@@ -446,6 +454,172 @@ connection_edge_end(edge_connection_t *conn, uint8_t reason)
   return 0;
 }
 
+/**
+ * Add a half-closed connection to the list, to watch for activity.
+ *
+ * These connections are removed from the list upon receiving an end
+ * cell.
+ */
+static void
+connection_half_edge_add(edge_connection_t *conn,
+                         origin_circuit_t *circ)
+{
+  half_edge_t *half_conn = tor_malloc_zero(sizeof(half_edge_t));
+
+  if (!circ->half_streams) {
+    circ->half_streams = smartlist_new();
+  }
+
+  half_conn->stream_id = conn->stream_id;
+
+  // How many sendme's should I expect?
+  half_conn->sendmes_pending =
+   (STREAMWINDOW_START-conn->package_window)/STREAMWINDOW_INCREMENT;
+
+   // Is there a connected cell pending?
+  half_conn->connected_pending = conn->base_.state ==
+      AP_CONN_STATE_CONNECT_WAIT;
+
+  /* Data should only arrive if we're not waiting on a resolved.
+   * It can arrive after waiting on connected, because of optimistic
+   * data. */
+  if (conn->base_.state != AP_CONN_STATE_RESOLVE_WAIT) {
+    // How many more data cells can arrive on this id?
+    half_conn->data_pending = conn->deliver_window;
+  }
+
+  smartlist_add(circ->half_streams, half_conn);
+}
+
+/**
+ * Check if this stream_id is in a half-closed state. If so,
+ * check if it still has data cells pending, and decrement that
+ * window if so.
```

Comment:
I like the idea of factoring out the search into a separate function. I did this in 8257aab4e254e3aa2ad442456abe9b4d7c51ff95.

It seems safer to me for callers to use explicit named functions for each of the window checks, rather than making them pass in a pointer to an int or something. This way, caller misuse is more easily spotted, and we don't have to worry about dangling pointers.

In my mind, it makes sense to have these calls be questions rather than action statements. This is consistent with the philosophy I used for the patch. None of these functions perform any actions that alter behavior. They only tell the caller if the cell counts as valid data for the control port statistics, so that the caller can count it (or not).

Reply:
(e08e36f80af86482a94ce9deea05b625643ae9a5 has an additional fixup to make the search function take a const list).

Reply:
Thanks, the refactored code looks much better.